src/irule.cpp

```diff
@@ -322,11 +325,25 @@ main( int argc, char **argv ) {
                 if ( rulegen ) {
                     if ( startsWith( buf, "INPUT" ) || startsWith( buf, "input" ) ) {
                         gotRule = 1;
-                        trimSpaces( trimPrefix( buf ) );
+                        constexpr auto skip_len{5};
+                        const auto copy_length{std::strlen(buf) + 1 - skip_len};
+
+                        // Blank out "INPUT"
+                        std::memmove(buf, buf + skip_len, copy_length);
+
+                        auto res{boost::algorithm::trim_left_copy(std::string{buf})};
+                        std::strncpy(buf, res.c_str(), res.size() + 1);
```

Comment:
I'm a little confused by these lines. Please explain what's happening.

Reply:
The top line:

It removes, in this case, "INPUT" from `buf` in place. You can see this also used in `trimPrefix(...)`.

We know it'll remove "INPUT" because of how `startsWith(...)` behaves:

https://github.com/irods/irods/blob/cf4744ecd539f3ea3847f957d8d1121a468100d9/lib/core/src/rcMisc.cpp#L4449-L4451

Relying on its behavior, we can "hard-code" the number of chars to skip (this is `skip_len`).

---

The bottom two lines:

`boost::algorithm::trim_left_copy(...)` is templated, so its input is its return type. This does not work with c-strings. So, `buf` gets converted to a `std::string`, is trimmed, and returned in `res`.

We then copy the trimmed `res` into `buf`, so that the trimmed result is available in the loop.

Reply:
Ok. According to cppreference, the call to `std::memmove` will essentially shift everything to the left by `skip_len` characters. That will result in `INPUT` being removed. That also means L334 and L335 aren't necessary.

See https://gcc.godbolt.org/z/xPv6bEhqd.

Am I missing a detail somewhere?

Reply:
Yes, on `memmove`, it is for removing the matched string.

---

For missing details,

The original behavior was to both clear out the matched string, and clear out spaces afterwards. This behavior can be seen in both `trimPrefix(...)` and `trimSpaces(...)`, which were both originally called on `buf`. This should be visible in the removed code.

I don't know enough about rules to say if this is necessary, so replicating the original behavior was what I went for.

Reply:
That's the correct approach.

As @trel has said, lean on the test suite. If things pass, then we can be confident we haven't gone backwards. The test suite allows us to experiment and has done a very good job at telling us when we've done too much.

Reply:
Resolving.